webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/policies/RuleTarget.scala

```diff
@@ -295,17 +295,20 @@ object RuleTarget extends Loggable {
       targets:          Set[RuleTarget],
       allNodes:         Map[NodeId, Boolean /* isPolicyServer */ ],
       groups:           Map[NodeGroupId, Chunk[NodeId]],
-      allNodesAreThere: Boolean = true // if we are working on a subset of node, set to false
   ): Chunk[NodeId] = {
-    val result = getNodeIdsChunkRec(Chunk.fromIterable(targets), allNodes, groups, allNodesAreThere)
-    result.distinct
+    val result = getNodeIdsChunkRec(Chunk.fromIterable(targets), allNodes.toMap, groups)
+
+    // we always must intersect with the keySet of nodes, because sometime node ids in groups is a super set
+    // of the given allNodes, for ex!
+    // - for static group where some of the nodes were deleted,
+    // - when we have tenants (Rudder 8.1+=
+    Chunk.fromIterable(result.toSet.intersect(allNodes.keySet))
```

Comment:
if we are converting to a set at the end, wouldn't it be cheaper to deal with a set all along ?

Reply:
it's not the case in latter branches, so if we change here, we will have to change in upmerge. 7.3 
I can do it for clarity on that PR, but it's not what we will get in 8.0/8.1